### PR TITLE
[Perf][GLM-5.3-Flash] Dense/masked-MHA sparse prefill for the NoPE (256, 0, 256) layout + skip the NoPE K concat

### PR DESCRIPTION
**Pull request by JaredforReal**
## Purpose

One of three independent GLM-5.3-Flash perf PRs 

On main, GLM-5.3-Flash's MLA layout (qk_nope 256, qk_rope 0, v 256) is unknown to the FlashAttention prefill backend and to the masked-MHA allow-list, so `MLACommonImpl` logs `No MLA prefill backend supports this model` and **every prefill token goes through the per-token top-k MQA kernel**. Profiling shows that kernel is KV-gather bound at the HBM roofline: 5.9 ms/layer per 16k-token chunk, and still 1.6 ms/layer for 8x2048 where dense MHA needs ~0.2 ms.

Three commits:
1. **Register the (256, 0, 256) dims with `FlashAttnPrefillBackend`** (7 lines). Same lines as in the SM90 wiring PR vllm-project/vllm#55385; kept separate so it can be dropped once that lands. This alone gives the dense-MHA prefill path for sequences ≤ index_topk (2048).
2. **Masked-MHA allow-list**: `_is_masked_mha_available` only knew (128, 512, 128, 64, 128) and (64, 512, 192, 64, 256); GLM-5.3-Flash is (64, 512, 256, 0, 256). The kernels are identical for qk_head_dim 256 / v 256, so sequences between 2k and the existing FLASHINFER_MLA_SPARSE thresholds (36k at TP4, 64k at TP8) now use FA4 masked MHA instead of the per-token top-k MQA kernel.
3. **NoPE K concat**: `_concat_k_nope_k_pe` returns `k_nope` directly when there is no RoPE part instead of allocating and copying a same-sized tensor (134 MB/layer per 16k chunk).

## Test Plan

- Unit / kernel tests listed under "Tests" below.
- Serving benchmark, all builds back-to-back in one session on 4x GB300 (SM100), TP4, prefix caching disabled:
  ```bash
  vllm serve zai-org/GLM-5.3-Flash -tp 4 --attention-backend FLASHINFER_MLA_SPARSE \
    --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384 \
    --gpu-memory-utilization 0.9 --no-enable-prefix-caching
  vllm bench serve --backend vllm --model glm53-flash --dataset-name random --random-range-ratio 0 \
    --ignore-eos --seed 0 --random-input-len <in> --random-output-len <out> --max-concurrency <c> \
    --num-prompts <n> --num-warmups <c> --percentile-metrics ttft,tpot,itl,e2el
  ```
 0 prefix-cache hit 
- Accuracy per build: `lm_eval --model local-completions --tasks gsm8k --num_fewshot 5 --gen_kwargs temperature=0` (1319 questions)

### Performance of this PR alone

Measured on top of #55736 + #55737 (main + #55736 + #55737) vs the same + this PR, 4x GB300, TP4, `--attention-backend FLASHINFER_MLA_SPARSE --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384`, **prefix caching disabled** (hit rate 0.0% checked in the server log), `vllm bench serve` random dataset with warmups, back-to-back in the same session. Decode tok/s is the steady-state window value; TPOT is the per-request median.

| point | metric | +#55736 +#55737 | +#55736 +#55737 +#55738 | delta |
|---|---|---|---|---|
| prefill 8x2048 | mean TTFT ms | 395 | 361 | -8.5% |
| prefill 2x8192 | mean TTFT ms | 330 | 315 | -4.6% |
| prefill 8x32768 | mean TTFT ms | 4693 | 4347 | -7.4% |
| prefill 2x65536 | mean TTFT ms | 2450 | 2365 | -3.5% |
| 32k/256 c=16 | decode tok/s | 316 | 322 | +1.9% |
| 32k/256 c=16 | mean TTFT ms | 3642 | 3482 | -4.4% |
| 32k/256 c=16 | median TPOT ms | 37.50 | 36.97 | -1.4% |
| 1k/512 c=1 | decode tok/s | 147 | 147 | -0.2% |
| 1k/512 c=1 | median TPOT ms | 6.47 | 6.49 | +0.3% |
| 1k/512 c=64 | decode tok/s | 3132 | 3133 | +0.0% |
| 1k/512 c=64 | median TPOT ms | 19.24 | 19.09 | -0.8% |
| 1k/512 c=256 | decode tok/s | 7717 | 7755 | +0.5% |
| 1k/512 c=256 | median TPOT ms | 31.30 | 31.29 | -0.0% |

Prefill up to the masked-MHA threshold (≤ 36k at TP4): −8.5% TTFT for 8x2048 (dense MHA), −7.4% for 8x32k (masked MHA), −4.4% TTFT in the 32k-context serving point. 2x8192 shows −4.6%, but that 16-request point has a ~4% run-to-run noise floor (last column of the ablation below), so treat it as indicative only. 64k prompts are above the threshold and stay on the sparse kernel; the −3.5% there (0.0% noise floor) is the K-concat removal alone. Decode is unchanged (±0.8%).

### Accuracy (per build, same session)

Each rung of the ablation was also evaluated on its own: **gsm8k** (1319 questions, 5-shot, greedy, lm_eval `local-completions`), **prompt-logprob agreement** on real 4k/12k/30k/60k prompts (mean |Δ logprob| per token vs the main build; the "main again" row is the run-to-run noise floor of this FP8 model), and a **needle-in-a-haystack** retrieval set (12 codes per length at 6k/16k/30k tokens, greedy, `reasoning_effort=low`), which exercises the 2k-36k prefill range where #55738 switches kernels.

| build | gsm8k flexible-extract % | gsm8k strict-match % |
|---|---|---|
| main | 93.33 ± 0.69 | 93.10 ± 0.70 |
| +#55736 +#55737 +#55738 | 93.10 ± 0.70 | 93.10 ± 0.70 |

### Caveat: runtime JIT

The first request in each new sequence-length bucket JIT-compiles the FA4 CuTeDSL `mask_mod` variant (`BlackwellFusedMultiHeadAttentionForward`) plus the Triton `_scatter_topk_single_req_kernel` (~7 s each; `jit_monitor` warns). `FA4MLAPrefillKernel.get_warmup_keys` does not cover mask_mod, so this is a pre-existing gap for every masked-MHA model; a warmup for the masked variants should be a follow-up before relying on this in production.

## Duplicate-work check

`gh pr list --repo vllm-project/vllm --state open --search "NoPE masked MHA 256"` / `"GLM-5.3-Flash prefill"`: #55385 covers the FA dims (commit 1 here is the same 7 lines); #54951 / #55222 / #55543 touch the indexer prefill sharding, the indexer workspace and SM90 fp8 KV. None adds the masked-MHA layout or the K-concat skip.
- vllm-project/vllm#53969 / #55277 add NoPE support to the SM120 FlashInfer sparse backend only (`flashinfer_mla_sparse_sm120.py`); no overlap with the masked-MHA allow-list or the K-concat skip.

## Tests

- New unit tests (all pass on GB300, 14 cases):
  - `tests/v1/attention/test_sparse_mla_backends.py::test_sparse_backend_prefill_correctness[glm53_flash_nope_hd256_v256-*]`: dense MHA, masked MHA and masked MHA with chunked context for the NoPE `(64, 256, 0, 256)` layout against the per-token SDPA reference.
  - `tests/v1/attention/test_sparse_mla_backends.py::test_is_masked_mha_available_model_dims`: allow-list accepts the DeepSeek-V3.2, GLM-5 and NoPE GLM-5.3-Flash geometries and rejects quantized KV, FA3, `(256, 64, 256)` and a different head count.
  - `tests/v1/attention/test_mla_prefill_selector.py::TestBackendValidation::test_flash_attn_accepts_glm53_flash_nope_dimensions`: `FlashAttnPrefillBackend.validate_configuration` accepts `(256, 0, 256)` and still rejects `(256, 64, 256)`.
  - `tests/v1/attention/test_mla_backends.py::test_concat_k_nope_k_pe_matches_torch_cat[rope|nope]`: `_concat_k_nope_k_pe` equals `torch.cat` of `k_nope` and the broadcast `k_pe`; with no RoPE part it returns `k_nope` itself (same `data_ptr`).
- `tests/kernels/test_mhc_kernels.py`, `tests/kernels/test_glm5next_kda_recurrent_strided.py` unaffected; prompt-logprob A/B on real 4k/12k/30k/60k prompts vs main: mean|Δ| 0.06-0.12 vs a main-vs-main noise floor of 0.04-0.11 (≈2.5x at 4k where dense/masked MHA replaces the sparse kernel), identical next tokens; **gsm8k 1319 questions (5-shot, greedy): main 93.03% ± 0.70, full series 92.65% ± 0.72** (measured on the three PRs together).
- `pre-commit run ruff-check / ruff-format`: passed.
- AI assistance: developed with Claude Fable 5.1 and reviewed by the submitter; profiling notes and raw benchmark logs available on request.

## Notes for review
AI assistance: developed with Claude and reviewed by the submitter.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/mla/prefill/flash_attn.py [hunk 1]
```diff
@@ -327,6 +327,13 @@ def supports_mla_dimensions(cls, mla_dimensions: MLADimensions) -> bool:
                 qk_rope_head_dim=64,
                 v_head_dim=128,
             ),
+            # GLM-5.x NoPE layout: qk_head_dim 256 + 0 and v_head_dim 256 run
```
Comment (ZJY0516) [inline vllm/v1/attention/backends/mla/prefill/flash_attn.py]:
```suggestion
            # GLM-5.3-Flash NoPE layout: qk_head_dim 256 + 0 and v_head_dim 256 run
```

**Review by ZJY0516 — APPROVED** [review]

**Comment by ZJY0516** [comment]
@JaredforReal please add some unit tests, the coplilot review is right
